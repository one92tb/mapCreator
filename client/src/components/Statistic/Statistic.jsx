import React, { Component } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { fetchSelectedMarkers } from "../../actions/fetchSelectedMarkers";
import { fetchRecords } from "../../actions/fetchRecords";
import BarGraph from "./BarGraph/BarGraph";
import PieGraph from "./PieChart/PieGraph";
import { Container, Row, Col } from "reactstrap";

const ContainerStyle = styled.div`
  height: 40%;
  padding: 0 !important;
`;

const RowStyle = styled.div`
  height: 100%;
  margin: 0 !important;
`;
const ColStyle = styled.div`
  padding: 0!important
  height: 100%;
`;

const Wrapper = styled.div`
  height: 100%;
  padding: 40px 20px;
  background: #f2f2f2;
`;

const Form = styled.form`
  height: calc(10% - 20px);
  display: flex;
  padding-right: 30px;
  margin-bottom: 20px;
  justify-content: flex-end;
`;
const Input = styled.input`
  height: 40px;
  width: 250px;
  text-align: center;
  border-radius: 5px;
  border: 1px solid #bfbfbf;
`;

class Statistic extends Component {
  constructor(props) {
    super(props);
    this.state = {
      city: ""
    };
  }

  componentDidMount() {
    this.props.fetchSelectedMarkers();
    this.props.fetchRecords();
  }

  handleChange = e => {
    this.setState({
      city: e.target.value
    });
  };

  render() {
    console.log(this.state);
    const displayMarkers = Object.entries(
      this.props.selectedMarkers
        .filter(marker => {
          if (
            this.state.city === "" ||
            marker.city.toLowerCase().search(this.state.city.toLowerCase()) !==
              -1
          ) {
            return marker;
          }
        })
        .reduce((obj, el, id) => {
          obj[el.name] = obj[el.name] ? ++obj[el.name] : 1;
          return obj;
        }, {})
    );
    const displaySumMarkers = [["markers", this.props.selectedMarkers.length]];

    return (
      <Wrapper>
        <Form>
          <label>
            <Input
              onChange={this.handleChange}
              type="text"
              name="city"
              placeholder="search your city"
            />
          </label>
        </Form>
        <BarGraph displayMarkers={displayMarkers} />
        <Container fluid tag={ContainerStyle}>
          <Row tag={RowStyle}>
            <Col lg="6" tag={ColStyle}>
              <PieGraph displayMarkers={displaySumMarkers} />
            </Col>
            <Col lg="6" tag={ColStyle}>
              <PieGraph displayMarkers={displayMarkers} />
            </Col>
          </Row>
        </Container>
      </Wrapper>
    );
  }
}

const mapStateToProps = state => ({
  selectedMarkers: state.selectedMarker.selectedMarkers,
  markers: state.marker.records
});

const mapDispatchToProps = {
  fetchSelectedMarkers,
  fetchRecords
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Statistic);

/*{Object.entries(
  state.selectedMarker.selectedMarkers.reduce((obj, el, id) => {
    obj[el.name] = obj[el.name] ? ++obj[el.name] : 1;
    return obj;
  }, {})
).map(el => {
  console.log(el);
  return {
    name: el[0],
    value: el[1]
  };
})}*/
