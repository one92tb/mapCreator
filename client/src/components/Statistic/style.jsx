import styled from "styled-components";

const ContainerStyle = styled.div`
  height: 40%;
  padding: 0 !important;
`;

const RowStyle = styled.div`
  height: 100%;
  margin: 0 !important;
`;
const ColStyle = styled.div`
  padding: 0!important
  height: 100%;
`;

const Wrapper = styled.div`
  height: 100%;
  padding: 40px 20px;
  background: #f2f2f2;
`;

const Form = styled.form`
  height: calc(10% - 20px);
  display: flex;
  padding-right: 30px;
  margin-bottom: 20px;
  justify-content: flex-end;
`;
const Input = styled.input`
  height: 40px;
  width: 250px;
  text-align: center;
  border-radius: 5px;
  border: 1px solid #bfbfbf;
`;

export { ContainerStyle, RowStyle, ColStyle, Wrapper, Form, Input };
