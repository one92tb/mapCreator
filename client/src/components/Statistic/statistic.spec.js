import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import configureStore from "redux-mock-store";
import { Statistic } from "./Statistic";

describe("statistic component", () => {
  let wrapper;

  it("passess all props to Statistic component", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      }
    ];

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it("componentDidMount", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      }
    ];

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );

    expect(fetchIndicators).toHaveBeenCalledTimes(1);
    expect(fetchMarkers).toHaveBeenCalledTimes(1);
  });

  it("handle change city", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 2,
        lat: 50.92841093800815,
        lng: 15.640584390869208,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "123123213.png",
        id: 4,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];

    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );

    const input = wrapper.find("input");
    input.simulate("change", { target: { value: "Jelenia Góra" } });
    expect(input.props().name).toEqual("city");
    expect(wrapper.state().city).toEqual("Jelenia Góra");
  });

  it("sumEachIndicator method for Jelenia Góra", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "123123213.png",
        id: 4,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );

    const { city } = wrapper.instance().state;
    const expectedResult = [["test1", 1]];

    wrapper.setState({ city: "Jelenia Góra" });
    expect(wrapper.instance().sumEachIndicator()).toEqual(expectedResult);
  });

  it("sumEachIndicator method for all indicators", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];
    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "123123213.png",
        id: 4,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "12323213.png",
        id: 5,
        lat: 50.9332387715648635,
        lng: 15.832054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );
    wrapper.setState({ city: "" });
    const { city } = wrapper.instance().state;

    const result1 = indicators.filter((indicator, id, arr) => {
      return (
        (city === "" ||
          indicator.city.toLowerCase().search(city.toLowerCase()) !== -1) &&
        indicator
      );
    });

    const expectedResult = [["test1", 1], ["test2", 2]];
    expect(wrapper.instance().sumEachIndicator()).toEqual(expectedResult);
    expect(wrapper.instance().sumEachIndicator()[0][1]).toEqual(1);
  });

  it("sumAllIndiacators method", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const indicators = [
      {
        city: "Jelenia Góra 58-500",
        country: "Polska",
        icon: "1548784752159.png",
        id: 3,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test1",
        street: "Unnamed Road",
        userId: 1
      },
      {
        city: "Wrocław",
        country: "Polska",
        icon: "123123213.png",
        id: 4,
        lat: 50.933387715648635,
        lng: 15.840054910888739,
        name: "test2",
        street: "Unnamed Road",
        userId: 1
      }
    ];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );
    const expectedResult = [["All markers", 2]];
    expect(wrapper.instance().sumAllIndiacators()).toEqual(expectedResult);
  });

  it("render Graphs depends on indicators length - should return Graphs", () => {
    const markers = [
      {
        icon: "1548784752159.png",
        id: 1,
        name: "test1",
        userId: 1
      },
      {
        icon: "1550173679016.png",
        id: 2,
        name: "basen",
        userId: 1
      }
    ];

    const indicators = [];
    const fetchIndicators = jest.fn();
    const fetchMarkers = jest.fn();

    wrapper = shallow(
      <Statistic
        markers={markers}
        indicators={indicators}
        fetchMarkers={fetchMarkers}
        fetchIndicators={fetchIndicators}
      />
    );

    expect(
      wrapper
        .children()
        .at(1)
        .get(0).props.children.props.children
    ).toEqual("You have not any data to display on the charts");
  });
});
