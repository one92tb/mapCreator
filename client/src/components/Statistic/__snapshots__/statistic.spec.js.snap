// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`statistic component passess all props to Statistic component 1`] = `
<div>
  <form>
    <label>
      <input
        name="city"
        onChange={[Function]}
        placeholder="search your city"
        type="text"
      />
    </label>
  </form>
  <Component
    displayMarkers={[Function]}
  />
  <div
    fluid={true}
    tag={[Function]}
  >
    <div
      tag={[Function]}
    >
      <div
        lg="12"
        tag={[Function]}
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xl="6"
      >
        <div>
          <PieGraph
            displayMarkers={[Function]}
          />
        </div>
      </div>
      <div
        lg="12"
        tag={[Function]}
        widths={
          Array [
            "xs",
            "sm",
            "md",
            "lg",
            "xl",
          ]
        }
        xl="6"
      >
        <div>
          <PieGraph
            displayMarkers={[Function]}
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
